fuse_models/src/odometry_2d_publisher.cpp

```diff
@@ -63,31 +64,65 @@ namespace fuse_models
 Odometry2DPublisher::Odometry2DPublisher() :
   fuse_core::AsyncPublisher(1),
   device_id_(fuse_core::uuid::NIL),
+  logger_(rclcpp::get_logger("uninitialized")),
   latest_stamp_(rclcpp::Time(0, 0, RCL_ROS_TIME)),
-  latest_covariance_stamp_(rclcpp::Time(0, 0, RCL_ROS_TIME)),
-  publish_timer_spinner_(1, &publish_timer_callback_queue_)
+  latest_covariance_stamp_(rclcpp::Time(0, 0, RCL_ROS_TIME))
 {
 }
 
+void Odometry2DPublisher::initialize(
+  fuse_core::node_interfaces::NodeInterfaces<ALL_FUSE_CORE_NODE_INTERFACES> interfaces,
+  const std::string & name)
+{
+  interfaces_ = interfaces;
+  fuse_core::AsyncPublisher::initialize(interfaces, name);
+}
+
 void Odometry2DPublisher::onInit()
 {
+  logger_ = interfaces_.get_node_logging_interface()->get_logger();
+  clock_ = interfaces_.get_node_clock_interface()->get_clock();
+
+  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(interfaces_);
+
   // Read settings from the parameter sever
-  device_id_ = fuse_variables::loadDeviceId(private_node_handle_);
+  device_id_ = fuse_variables::loadDeviceId(interfaces_);
 
-  params_.loadFromROS(private_node_handle_);
+  params_.loadFromROS(interfaces_, name_);
 
   if (!params_.invert_tf && params_.world_frame_id == params_.map_frame_id)
   {
-    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(params_.tf_cache_time);
-    tf_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf_buffer_, node_handle_);
+    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(
+      clock_,
+      params_.tf_cache_time.to_chrono<std::chrono::nanoseconds>()
+      // , interfaces_  // NOTE(methylDragon): This one is pending a change on tf2_ros/buffer.h
+                        // TODO(methylDragon): See above ^
+    );
+
+    tf_listener_ = std::make_unique<tf2_ros::TransformListener>(
+      *tf_buffer_,
+      interfaces_.get_node_base_interface(),
+      interfaces_.get_node_logging_interface(),
+      interfaces_.get_node_parameters_interface(),
+      interfaces_.get_node_topics_interface());
   }
 
-  odom_pub_ = node_handle_.advertise<nav_msgs::Odometry>(ros::names::resolve(params_.topic), params_.queue_size);
-  acceleration_pub_ = node_handle_.advertise<geometry_msgs::AccelWithCovarianceStamped>(
-      ros::names::resolve(params_.acceleration_topic), params_.queue_size);
-
-  publish_timer_node_handle_.setCallbackQueue(&publish_timer_callback_queue_);
-  publish_timer_spinner_.start();
+  // Advertise the topics
+  rclcpp::PublisherOptions pub_options;
+  pub_options.callback_group = cb_group_;
+
+  // NOTE(CH3): We might need to prepend the name here like with the other publishers
+  //            ... But will that cause the subscriptions on the models to fail???
```

Comment:
What name might need to be prepended? I see `jointTopicName` is already used below.

Reply:
Ah, this comment was about using that `joinTopicName` call. I had mistakenly thought that the original ROS 1 code didn't use the same sort of private node handle namespacing like the other publishers. Sorry for the noise :grimacing: